openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxNamedDestinationManager.java

```diff
@@ -0,0 +1,73 @@
+package com.openhtmltopdf.pdfboxout;
+
+import com.openhtmltopdf.css.constants.CSSName;
+import com.openhtmltopdf.css.constants.IdentValue;
+import com.openhtmltopdf.layout.SharedContext;
+import com.openhtmltopdf.render.Box;
+import com.openhtmltopdf.render.RenderingContext;
+import com.openhtmltopdf.util.LogMessageId;
+import com.openhtmltopdf.util.XRLog;
+import org.apache.pdfbox.cos.COSDictionary;
+import org.apache.pdfbox.pdmodel.PDDestinationNameTreeNode;
+import org.apache.pdfbox.pdmodel.PDDocument;
+import org.apache.pdfbox.pdmodel.PDDocumentNameDictionary;
+import org.apache.pdfbox.pdmodel.interactive.documentnavigation.destination.PDPageDestination;
+import org.apache.pdfbox.pdmodel.interactive.documentnavigation.destination.PDPageXYZDestination;
+
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.logging.Level;
+
+public class PdfBoxNamedDestinationManager {
+
+    private final PDDocument _writer;
+    private final SharedContext _sharedContext;
+    private final float _dotsPerPoint;
+    private final PdfBoxFastOutputDevice _od;
+
+    public PdfBoxNamedDestinationManager(PDDocument doc, SharedContext ctx, float dotsPerPoint, PdfBoxFastOutputDevice od) {
+        this._writer = doc;
+        this._sharedContext = ctx;
+        this._dotsPerPoint = dotsPerPoint;
+        this._od = od;
+    }
+
+    public void processNamedDestinations(RenderingContext c, Box root) {
+        Map<String, Box> idMap = _sharedContext.getIdMap();
+        if (idMap != null && !idMap.isEmpty()) {
+            Map<String, PDPageDestination> names = new LinkedHashMap<>();
+            idMap.forEach((id, box) -> {
+                if (box.getStyle().isIdent(CSSName.FS_NAMED_DESTINATION, IdentValue.CREATE)) {
+                    names.put(id, createDestination(c, box, root));
+                }
+            });
```

Comment:
I'm a little concerned about performing this linear search every time.

Ideally, I think while the `Box`s are being created, or while the stylesheet is being parsed, we could mark a flag in  `SharedContext`, notifying the `PdfBoxNamedDestinationManager` that this feature is in use. Then, we could ignore the linear search for most PDFs that don't use named destinations.

I looked around for a while and the best option I found was to set the flag inside `CSSParser#declaration`:
```java
                    if (valid) {
+                       if (cssName == CSSName.FS_NAMED_DESTINATION && values.stream().anyMatch(propertyValue -> propertyValue.getIdentValue() == IdentValue.CREATE)) {
+                           ThreadCtx.get().sharedContext().setUsingFsNamedDestination(true);
+                       }
+
                        try {
                            PropertyBuilder builder = CSSName.getPropertyBuilder(cssName);
                            ruleset.addAllProperties(builder.buildDeclarations(
                                    cssName, values, ruleset.getOrigin(), important, !inFontFace));

```

Reply:
No problem, done.